## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

/var/folders/gw/s1_67f755hld2_g983ssmt2w0000gn/T/ipykernel_19852/2050583822.py:14: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
contrib.shape

(1125659, 18)

- **1a answer here:** The file contains 18 columns as outlined in the documentation, having 1.1M rows seems reasonable as the data shows all the contribution related transactions. 

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
list(contrib.columns)

['cmte_id',
 'cand_id',
 'cand_nm',
 'contbr_nm',
 'contbr_city',
 'contbr_st',
 'contbr_zip',
 'contbr_employer',
 'contbr_occupation',
 'contb_receipt_amt',
 'contb_receipt_dt',
 'receipt_desc',
 'memo_cd',
 'memo_text',
 'form_tp',
 'file_num',
 'tran_id',
 'election_tp']

- **1b answer here:** Yes. All 18 columns are defined in the documentation.

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
contrib.head()

cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26          NaN       X   
1            RETIRED             200.00       2016-04-20          NaN       X   
2            RETIRED               5.00       2016-04-02          NaN       X   
3  SOFTWARE ENGINEER              40.00       2016-03-04          NaN     NaN   
4         PHARMACIST              35.00       2016-03-05          NaN     NaN   

                             memo_text form_tp  file_num      tran_id  \
0               * HILLARY VICTORY FUND    SA18   1091718     C4768722   
1               * HILLARY VICTORY FUND    SA18   1091718     C4747242   
2               * HILLARY VICTORY FUND    SA18   1091718     C4666603   
3  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWA097   
4  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   

  election_tp  
0       P2016  
1       P2016  
2       P2016  
3       P2016  
4       P2016

- **1c answer here:** cand_id is a number that represents each candidate and the candidate name is outlined under cand_nm which appears as a legal name shown in this format "Last name, First name" and the case of Hillary Clinton it also includes her "Rodham" name as well. contr_st stands for contribution state and it appears as *CA* for California. 

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
election_type_cnt = contrib.pivot_table(index=['election_tp'], values=['contb_receipt_amt'], aggfunc={'contb_receipt_amt': ['sum']})

election_type_cnt

contb_receipt_amt
                          sum
election_tp                  
G2016           43,325,812.02
P2016           93,350,485.44
P2020               15,000.00

- **1d answer here:** The documentation tells us that G is for general election and P is for Primary election. It appears that this letter is followed by the election year. Since this is historical information on the 2016 election it makes sense that the total contrinution amounts for both the General (G2016) and Primary elections (P2016) are the largest values and the Primary 2020 (P2020) contributions seem more like outlier rows that we do not need in our analysis as we are focusing on 2016.

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
contrib.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** contbr_zip is object or string, contb_receipt_amt is float64 (64-bit floating-point number), and contb_receipt_dt is datetime64[ns] which means it is stored with datetime information with nanosecond precision.

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
null_count = contrib.isnull().sum()
print(null_count)

most_non_nulls_column = contrib.count().idxmax()

print(f"The column with the most non-null values is: {most_non_nulls_column}")

cmte_id                    0
cand_id                    0
cand_nm                    0
contbr_nm                  0
contbr_city               26
contbr_st                  0
contbr_zip                95
contbr_employer       157902
contbr_occupation      10399
contb_receipt_amt          0
contb_receipt_dt           0
receipt_desc         1110614
memo_cd               981391
memo_text             624511
form_tp                    0
file_num                   0
tran_id                    0
election_tp             1425
dtype: int64
The column with the most non-null values is: cmte_id


- **1f answer here:** The column with the most non-null values is: cmte_id. The column receipt_desc has 1.1M nulls, given that the total dataset is 1,125,659 rows this column is 98.7% empty and thus it can easily be dropped as it is not adding value to the research. 

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE
candnm_match = contrib.pivot_table(index=['cand_id', 'cand_nm'], values=['tran_id'], aggfunc={'tran_id': ['count']})

# Resetting index to make 'cand_nm' a regular column
candnm_match = candnm_match.reset_index()

# Sorting the DataFrame based on the 'cand_nm' column
candnm_match = candnm_match.sort_values(by=('cand_nm', ''), ascending=True)

candnm_match

cand_id                    cand_nm tran_id
                                           count
16  P60008059                  Bush, Jeb    3130
7   P60005915        Carson, Benjamin S.   27362
18  P60008521   Christie, Christopher J.     333
0   P00003392    Clinton, Hillary Rodham  547211
9   P60006111  Cruz, Rafael Edward 'Ted'   57820
12  P60007242             Fiorina, Carly    4696
23  P80003379       Gilmore, James S III       3
15  P60007697         Graham, Lindsey O.     347
24  P80003478             Huckabee, Mike     531
17  P60008398              Jindal, Bobby      31
1   P20002671              Johnson, Gary    1591
6   P60003670            Kasich, John R.    3005
20  P60009685           Lessig, Lawrence     372
21  P60022654             McMullin, Evan     156
14  P60007671    O'Malley, Martin Joseph     397
13  P60007572          Pataki, George E.      20
5   P40003576                 Paul, Rand    4268
3   P20003281     Perry, James R. (Rick)     116
10  P60006723               Rubio, Marco   14092
11  P60007168           Sanders, Bernard  407172
2   P20002721       Santorum, Richard J.      88
4   P20003984                Stein, Jill    1907
22  P80001571           Trump, Donald J.   50168
8   P60006046              Walker, Scott     737
19  P60008885      Webb, James Henry Jr.     106

- **1g answer here:** after doing a count of tranactions by cand_id and cand_name it appears that the cand_id P60007168 is only assigned to Sanders, Bernard once, so no immediate issues spotted

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
state_cnt = contrib.pivot_table(index=['contbr_st'], values=['tran_id'], aggfunc={'tran_id': ['count']})

state_cnt

tran_id
             count
contbr_st         
CA         1125659

- **1h answer here:** All 1,125,659 transactions are from the state of California

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
transaction_id_cnt = contrib.pivot_table(index=['tran_id'], values=['contbr_st'], aggfunc={'contbr_st': ['count']})

# # Resetting index to make 'cand_nm' a regular column
transaction_id_cnt = transaction_id_cnt.reset_index()

# Finding the maximum count in the 'contbr_st' column to check if there are reasons to search further
max_count = transaction_id_cnt[('contbr_st', 'count')].max()

print("Max Count:", max_count)

# Filtering rows with count greater than 1
non_key_tran_ids = transaction_id_cnt[transaction_id_cnt[('contbr_st', 'count')] > 1]

non_key_tran_ids

Max Count: 4


tran_id contbr_st
                                 count
1068    A26C35A6066754130B99         3
1440    A340DF85B7F884133A20         3
2235    A4E50E2DD07E4475996F         3
2450    A5602AD777C8C4632B5A         4
3599    A7C22FA389E0348F98F0         3
...                      ...       ...
713610            SB28A.1286         2
713615            SB28A.1327         2
713975           SB28A.90929         2
713999            SB28A.9119         2
714002            SB28A.9127         2

[3419 rows x 2 columns]

- **1i answer here:** We cannot use tran_id as a key as the values are not unique, there are 3419 tran_ids that appear more than once

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE
negative_donations = contrib[contrib['contb_receipt_amt'] < 0]
negative_donations.head(20)

cmte_id    cand_id                    cand_nm               contbr_nm  \
19   C00574624  P60006111  Cruz, Rafael Edward 'Ted'        JOLLIFF, RICHARD   
23   C00574624  P60006111  Cruz, Rafael Edward 'Ted'        JOLLIFF, RICHARD   
81   C00574624  P60006111  Cruz, Rafael Edward 'Ted'        JOLLIFF, RICHARD   
190  C00574624  P60006111  Cruz, Rafael Edward 'Ted'        NOWELL, DIANA L.   
213  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
220  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
223  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
231  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
261  C00574624  P60006111  Cruz, Rafael Edward 'Ted'        NOWELL, DIANA L.   
307  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
312  C00574624  P60006111  Cruz, Rafael Edward 'Ted'      LICHTY, ANDREW MR.   
355  C00574624  P60006111  Cruz, Rafael Edward 'Ted'            MARQUIS, ZAN   
418  C00574624  P60006111  Cruz, Rafael Edward 'Ted'          NOWELL, ROBERT   
488  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   
492  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   
494  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   
498  C00574624  P60006111  Cruz, Rafael Edward 'Ted'          NOWELL, ROBERT   
583  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   
649  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   
654  C00574624  P60006111  Cruz, Rafael Edward 'Ted'  PALMER, RICHARD A. MR.   

                contbr_city contbr_st     contbr_zip  \
19                    CHICO        CA 959,289,507.00   
23                    CHICO        CA 959,289,507.00   
81                    CHICO        CA 959,289,507.00   
190  RANCHO SANTA MARGARITA        CA 926,884,928.00   
213               SAN DIEGO        CA 921,096,720.00   
220               SAN DIEGO        CA 921,096,720.00   
223               SAN DIEGO        CA 921,096,720.00   
231               SAN DIEGO        CA 921,096,720.00   
261  RANCHO SANTA MARGARITA        CA 926,884,928.00   
307               SAN DIEGO        CA 921,096,720.00   
312               SAN DIEGO        CA 921,096,720.00   
355             LOS ANGELES        CA 900,204,812.00   
418  RANCHO SANTA MARGARITA        CA 926,884,928.00   
488            LAGUNA WOODS        CA 926,372,766.00   
492            LAGUNA WOODS        CA 926,372,766.00   
494            LAGUNA WOODS        CA 926,372,766.00   
498  RANCHO SANTA MARGARITA        CA 926,884,928.00   
583            LAGUNA WOODS        CA 926,372,766.00   
649            LAGUNA WOODS        CA 926,372,766.00   
654            LAGUNA WOODS        CA 926,372,766.00   

                        contbr_employer  \
19                        SELF EMPLOYED   
23                        SELF EMPLOYED   
81                        SELF EMPLOYED   
190  CAPISTRAND UNIFIED SCHOOL DISTRICT   
213                       SELF EMPLOYED   
220                       SELF EMPLOYED   
223                       SELF EMPLOYED   
231                       SELF EMPLOYED   
261  CAPISTRAND UNIFIED SCHOOL DISTRICT   
307                       SELF EMPLOYED   
312                       SELF EMPLOYED   
355        MARQUIS PROPERTY COMPANY LTD   
418                                CUSD   
488                             RETIRED   
492                             RETIRED   
494                             RETIRED   
498                                CUSD   
583                             RETIRED   
649                             RETIRED   
654                             RETIRED   

                          contbr_occupation  contb_receipt_amt  \
19                                  RANCHER             -25.00   
23                                  RANCHER            -150.00   
81                                  RANCHER    

- **1j answer here:** According to the documentation negative donations are refunds to individuals and transfers from authorized committees. Looking at the first 20 rows in the df the transfer seems valid with the "Redesignation to General" memo_text

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE
min_date = contrib['contb_receipt_dt'].min()

max_date = contrib['contb_receipt_dt'].max()

print(f"the oldest date is: {min_date} and the most recent date is: {max_date}")

the oldest date is: 2013-11-05 00:00:00 and the most recent date is: 2016-10-19 00:00:00


- **1k answer here:** Yes, we can see that the data set includes a date in 2013 and a date in October of 2016. The dates are formatted correctly as YYYY-MM-DD which is expected for a format in the US.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:** No, we need to drop unnecessary columns and rows to narrow down to only the right timeframe and candidates (Bernie Sanders)

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:** The two questions our boss wants answered are:

Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
What day(s) of the month do most people donate?

Answer: *Yes*. we have columns for zip code, contribution amounts, and date information that we can use to determine an answer to these questions.

**1l.3** What columns are important?

- **1l.3 answer here:** 

cand_nm \
contbr_zip \
contb_receipt_amt \
contb_receipt_dt \
election_tp

**1l.4** What columns can be dropped?

- **1l.4 answer here:** These columns are completely unecessary at this time:

contbr_nm  \
contbr_employer \
contbr_occupation \
receipt_desc 


**1l.5** What are the data problems?

- **1l.5 answer here:** transaction id is not a unique key, dates are outside the timeframe specified, some donations are for 2020 based on the election_tp column 

**1l.6** List any assumptions so far:

- **1l.6 answer here:** 1.1M is a complete dataset, we were not told in the documentation how many rows to expect, election_tp is a letter followed by the year of the campaing that the contributors mean to donate to

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [13]:
# 2a YOUR CODE HERE
contrib = contrib[contrib['election_tp'] == 'P2016']

contrib.shape

(810481, 18)

**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [14]:
# 2b YOUR CODE HERE
contrib = contrib[contrib['cand_nm'] == 'Sanders, Bernard']

contrib.shape

(407171, 18)

**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [15]:
# Source of the data: California State Geoportal https://gis.data.ca.gov/datasets/CDEGIS::california-zip-codes/explore?location=37.148546%2C-119.005050%2C7.67

# Import CSV and inspect it. Note that PO_NAME is the city name in this df
zipcode_list = pd.read_csv('California_Zip_Codes.csv')
zipcode_list

OBJECTID  ZIP_CODE           PO_NAME STATE  POPULATION  POP_SQMI   SQMI  \
0           33     90001       Los Angeles    CA       59907 16,970.82   3.53   
1           34     90002       Los Angeles    CA       53573 17,281.61   3.10   
2           35     90003       Los Angeles    CA       71967 20,503.42   3.51   
3           36     90004       Los Angeles    CA       63987 20,508.65   3.12   
4           37     90005       Los Angeles    CA       41844 26,823.08   1.56   
...        ...       ...               ...   ...         ...       ...    ...   
1684      1717     96146    Olympic Valley    CA        1152    149.42   7.71   
1685      1718     96148       Tahoe Vista    CA        1487    991.33   1.50   
1686      1719     96150  South Lake Tahoe    CA       32999    151.87 217.28   
1687      1720     96161           Truckee    CA       19416     99.57 194.99   
1688      1721     96162           Truckee    CA         352    838.10   0.42   

      SHAPE_Length       SHAPE_Area  
0        44,694.66    98,280,502.07  
1        44,705.84    86,562,437.39  
2        52,236.41    97,919,865.81  
3        43,950.39    86,893,584.76  
4        43,301.53    43,528,162.40  
...            ...              ...  
1684     76,824.25   215,113,874.10  
1685     32,123.27    41,777,398.32  
1686    458,044.66 6,057,234,227.00  
1687    495,110.05 5,436,059,173.00  
1688     19,139.17    11,710,024.46  

[1689 rows x 9 columns]

In [16]:
# 2c YOUR CODE HERE
# Source of the data: California State Geoportal https://gis.data.ca.gov/datasets/CDEGIS::california-zip-codes/explore?location=37.148546%2C-119.005050%2C7.67

# Import CSV and inspect it. Note that PO_NAME is the city name in this df
zipcode_list = pd.read_csv('California_Zip_Codes.csv')
# zipcode_list

# Convert the 'ZIP_CODE' column to string in the 'zipcode_list' DataFrame
zipcode_list['ZIP_CODE'] = zipcode_list['ZIP_CODE'].astype(str)

# Extract the first 5 characters from 'contbr_zip' in the 'contrib' DataFrame
contrib['contbr_zip_first5'] = contrib['contbr_zip'].astype(str).str[:5]

# Add new Zipcode column to contrib df by doing an inner join on the first 5 digits of 'contbr_zip'
# and the 'ZIP_CODE' column in the 'zipcode_list' df
merged_df = pd.merge(contrib, zipcode_list[['ZIP_CODE']], left_on=['contbr_zip_first5'], right_on=['ZIP_CODE'], how='inner')

# Drop the temporary 'contbr_zip_first5' column
merged_df = merged_df.drop(columns=['contbr_zip_first5'])

# Print df
merged_df

cmte_id    cand_id           cand_nm         contbr_nm  \
0       C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE   
1       C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE   
2       C00577130  P60007168  Sanders, Bernard    MARTSOLF, MARK   
3       C00577130  P60007168  Sanders, Bernard     MCMAHON, JOHN   
4       C00577130  P60007168  Sanders, Bernard    ABBISS, ERIC C   
...           ...        ...               ...               ...   
397504  C00577130  P60007168  Sanders, Bernard     BAKER, ROBERT   
397505  C00577130  P60007168  Sanders, Bernard    FORBES, SANDRA   
397506  C00577130  P60007168  Sanders, Bernard   ZAMORA, GABRIEL   
397507  C00577130  P60007168  Sanders, Bernard     GOTTWALD, DAN   
397508  C00577130  P60007168  Sanders, Bernard     GOTTWALD, DAN   

          contbr_city contbr_st     contbr_zip           contbr_employer  \
0       REDONDO BEACH        CA 902,784,310.00  VERICOR ENTERPRISES INC.   
1       REDONDO BEACH        CA 902,784,310.00  VERICOR ENTERPRISES INC.   
2       REDONDO BEACH        CA 902,782,554.00           NORMAN LEAF, MD   
3       REDONDO BEACH        CA 902,785,313.00                      NONE   
4       REDONDO BEACH        CA 902,781,402.00                  TELETRAC   
...               ...       ...            ...                       ...   
397504         DUNLAP        CA 936,219,702.00                      NONE   
397505           ALTA        CA 957,010,607.00                   FTRADER   
397506         KERMAN        CA 936,300,125.00                      NONE   
397507        OAKLAND        CA 946,131,301.00                      SELF   
397508        OAKLAND        CA 946,131,301.00                      SELF   

         contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc  \
0               PHARMACIST              35.00       2016-03-05          NaN   
1               PHARMACIST             100.00       2016-03-06          NaN   
2       SURGICAL TECNICIAN              50.00       2016-03-04          NaN   
3                  RETIRED              50.00       2016-03-05          NaN   
4                 ENGINEER              50.00       2016-03-25          NaN   
...                    ...                ...              ...          ...   
397504        NOT EMPLOYED              50.00       2016-05-31          NaN   
397505        RETAIL SALES              15.00       2016-05-17          NaN   
397506        NOT EMPLOYED              27.00       2016-05-30          NaN   
397507              ARTIST               2.50       2016-05-23          NaN   
397508              ARTIST              25.00       2016-05-31          NaN   

       memo_cd                            memo_text form_tp  file_num  \
0          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
1          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
2          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
3          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
4          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
...        ...                                  ...     ...       ...   
397504     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397505     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397506     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397507     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397508     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   

            tran_id election_tp ZIP_CODE  
0       VPF7BKX3MB3       P2016    90278  
1       VPF7BKYBXV4       P2016    90278  
2       VPF7BKWTK79       P2016    90278  
3       VPF7BKX5V23       P2016    90278  
4       VPF7BMTR9Y9       P2016    90278  
...             ...         ...      ...  
397504  VPF7BPYDF08       P2016    93621  
397505  VPF7BPPXCF3       P2016    95701  
397506  VPF7BPWFWT3       P2016    93630  
397507  VPF7BPSZ3J9     

**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [17]:
# 2d YOUR CODE HERE

# Identify rows with negative 'contb_receipt_amt'
negative_donations_index = merged_df[merged_df['contb_receipt_amt'] < 0].index

# Drop rows with negative 'contb_receipt_amt'
merged_df = merged_df.drop(index=negative_donations_index)

# Display the modified DataFrame
merged_df

cmte_id    cand_id           cand_nm         contbr_nm  \
0       C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE   
1       C00577130  P60007168  Sanders, Bernard  LEONELLI, ODETTE   
2       C00577130  P60007168  Sanders, Bernard    MARTSOLF, MARK   
3       C00577130  P60007168  Sanders, Bernard     MCMAHON, JOHN   
4       C00577130  P60007168  Sanders, Bernard    ABBISS, ERIC C   
...           ...        ...               ...               ...   
397504  C00577130  P60007168  Sanders, Bernard     BAKER, ROBERT   
397505  C00577130  P60007168  Sanders, Bernard    FORBES, SANDRA   
397506  C00577130  P60007168  Sanders, Bernard   ZAMORA, GABRIEL   
397507  C00577130  P60007168  Sanders, Bernard     GOTTWALD, DAN   
397508  C00577130  P60007168  Sanders, Bernard     GOTTWALD, DAN   

          contbr_city contbr_st     contbr_zip           contbr_employer  \
0       REDONDO BEACH        CA 902,784,310.00  VERICOR ENTERPRISES INC.   
1       REDONDO BEACH        CA 902,784,310.00  VERICOR ENTERPRISES INC.   
2       REDONDO BEACH        CA 902,782,554.00           NORMAN LEAF, MD   
3       REDONDO BEACH        CA 902,785,313.00                      NONE   
4       REDONDO BEACH        CA 902,781,402.00                  TELETRAC   
...               ...       ...            ...                       ...   
397504         DUNLAP        CA 936,219,702.00                      NONE   
397505           ALTA        CA 957,010,607.00                   FTRADER   
397506         KERMAN        CA 936,300,125.00                      NONE   
397507        OAKLAND        CA 946,131,301.00                      SELF   
397508        OAKLAND        CA 946,131,301.00                      SELF   

         contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc  \
0               PHARMACIST              35.00       2016-03-05          NaN   
1               PHARMACIST             100.00       2016-03-06          NaN   
2       SURGICAL TECNICIAN              50.00       2016-03-04          NaN   
3                  RETIRED              50.00       2016-03-05          NaN   
4                 ENGINEER              50.00       2016-03-25          NaN   
...                    ...                ...              ...          ...   
397504        NOT EMPLOYED              50.00       2016-05-31          NaN   
397505        RETAIL SALES              15.00       2016-05-17          NaN   
397506        NOT EMPLOYED              27.00       2016-05-30          NaN   
397507              ARTIST               2.50       2016-05-23          NaN   
397508              ARTIST              25.00       2016-05-31          NaN   

       memo_cd                            memo_text form_tp  file_num  \
0          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
1          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
2          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
3          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
4          NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404   
...        ...                                  ...     ...       ...   
397504     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397505     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397506     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397507     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   
397508     NaN  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141   

            tran_id election_tp ZIP_CODE  
0       VPF7BKX3MB3       P2016    90278  
1       VPF7BKYBXV4       P2016    90278  
2       VPF7BKWTK79       P2016    90278  
3       VPF7BKX5V23       P2016    90278  
4       VPF7BMTR9Y9       P2016    90278  
...             ...         ...      ...  
397504  VPF7BPYDF08       P2016    93621  
397505  VPF7BPPXCF3       P2016    95701  
397506  VPF7BPWFWT3       P2016    93630  
397507  VPF7BPSZ3J9     

**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [18]:
# 2e YOUR CODE HERE
drop_columns = ['contbr_nm', 'contbr_employer', 'contbr_occupation', 'receipt_desc']

merged_df = merged_df.drop(columns=drop_columns)

merged_df

cmte_id    cand_id           cand_nm    contbr_city contbr_st  \
0       C00577130  P60007168  Sanders, Bernard  REDONDO BEACH        CA   
1       C00577130  P60007168  Sanders, Bernard  REDONDO BEACH        CA   
2       C00577130  P60007168  Sanders, Bernard  REDONDO BEACH        CA   
3       C00577130  P60007168  Sanders, Bernard  REDONDO BEACH        CA   
4       C00577130  P60007168  Sanders, Bernard  REDONDO BEACH        CA   
...           ...        ...               ...            ...       ...   
397504  C00577130  P60007168  Sanders, Bernard         DUNLAP        CA   
397505  C00577130  P60007168  Sanders, Bernard           ALTA        CA   
397506  C00577130  P60007168  Sanders, Bernard         KERMAN        CA   
397507  C00577130  P60007168  Sanders, Bernard        OAKLAND        CA   
397508  C00577130  P60007168  Sanders, Bernard        OAKLAND        CA   

           contbr_zip  contb_receipt_amt contb_receipt_dt memo_cd  \
0      902,784,310.00              35.00       2016-03-05     NaN   
1      902,784,310.00             100.00       2016-03-06     NaN   
2      902,782,554.00              50.00       2016-03-04     NaN   
3      902,785,313.00              50.00       2016-03-05     NaN   
4      902,781,402.00              50.00       2016-03-25     NaN   
...               ...                ...              ...     ...   
397504 936,219,702.00              50.00       2016-05-31     NaN   
397505 957,010,607.00              15.00       2016-05-17     NaN   
397506 936,300,125.00              27.00       2016-05-30     NaN   
397507 946,131,301.00               2.50       2016-05-23     NaN   
397508 946,131,301.00              25.00       2016-05-31     NaN   

                                  memo_text form_tp  file_num      tran_id  \
0       * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX3MB3   
1       * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKYBXV4   
2       * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKWTK79   
3       * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BKX5V23   
4       * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1077404  VPF7BMTR9Y9   
...                                     ...     ...       ...          ...   
397504  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141  VPF7BPYDF08   
397505  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141  VPF7BPPXCF3   
397506  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141  VPF7BPWFWT3   
397507  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141  VPF7BPSZ3J9   
397508  * EARMARKED CONTRIBUTION: SEE BELOW   SA17A   1094141  VPF7BPXJP52   

       election_tp ZIP_CODE  
0            P2016    90278  
1            P2016    90278  
2            P2016    90278  
3            P2016    90278  
4            P2016    90278  
...            ...      ...  
397504       P2016    93621  
397505       P2016    95701  
397506       P2016    93630  
397507       P2016    94613  
397508       P2016    94613  

[394484 rows x 15 columns]

- **2e answer here:** Dropped these unneccesary columns:

contbr_nm \
contbr_employer \
contbr_occupation \
receipt_desc

**2f.** List any assumptions that you made up to this point:

- **2f answer here:** a) 1.1M was a complete dataset, we were not told in the documentation how many rows to expect, election_tp is a letter followed by the year of the campaing that the contributors mean to donate to  b) 404,083 records sounds like a reasonable dataset for California

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [24]:
# 3a YOUR CODE HERE

# Create Pivot Table with count and sum columns
summary_table = merged_df.pivot_table(
    index=['ZIP_CODE'], 
    values=['contb_receipt_amt'], 
    aggfunc={'contb_receipt_amt': ['count', 'sum']}
)

# Rename columns for clarity
summary_table.columns = ['Count of donations', 'Total donations']

# Reset index
summary_table = summary_table.reset_index()

# Sort the summary table by "Total donations" in descending order
summary_table = summary_table.sort_values(by='Total donations', ascending=False)

# Display df
summary_table

ZIP_CODE  Count of donations  Total donations
820     94110                4046       294,061.13
823     94114                3043       208,602.37
826     94117                2884       201,545.43
935     94611                2516       150,232.12
1015    95060                2849       145,414.63
...       ...                 ...              ...
1360    95956                   1            20.00
1417    96071                   2            20.00
1292    95701                   1            15.00
1293    95703                   1            15.00
445     92365                   1            10.00

[1456 rows x 3 columns]

- **3a answer here:** zipcode 94110

**3b.** What day(s) of the month do most people donate? (10 points)

In [25]:
# 3b YOUR CODE HERE

# Extract day of the month
merged_df['day_of_month'] = merged_df['contb_receipt_dt'].dt.day

# Create Pivot Table with count and sum columns
date_table = merged_df.pivot_table(
    index=['day_of_month'], 
    values=['contb_receipt_amt'], 
    aggfunc={'contb_receipt_amt': ['count', 'sum']}
)

# Rename columns for clarity
date_table.columns = ['Count of donations', 'Total donations']

# Reset index
date_table = date_table.reset_index()

# Sort the summary table by "Total donations" in descending order
date_table = date_table.sort_values(by='Total donations', ascending=False)

# Display df
date_table

day_of_month  Count of donations  Total donations
28            29               23024     1,152,583.41
29            30               22285     1,018,179.12
30            31               18998     1,011,459.75
8              9               15740       901,382.44
13            14               17402       848,424.99
27            28               15837       813,656.13
9             10               13049       790,363.54
7              8               16370       762,003.32
26            27               12754       685,001.17
15            16               13886       674,212.57
1              2               10822       667,277.91
25            26               12248       619,366.82
17            18               11643       614,289.04
12            13               13490       614,241.88
16            17               12598       607,548.62
0              1               11613       601,209.56
11            12               10011       598,664.87
14            15               11729       595,949.17
22            23               11906       566,285.63
5              6               11306       557,671.35
19            20               10972       548,722.10
10            11               10159       547,329.22
3              4               11702       535,860.87
4              5               10287       525,296.73
20            21               10416       519,704.67
6              7               10215       507,817.40
24            25                9664       503,122.58
2              3                9268       474,593.69
18            19                8709       464,868.72
23            24                8012       441,447.13
21            22                8369       431,199.94

- **3b answer here:** the 29th of the month sees the most donations

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k